### PR TITLE
Fix chunking issues with BlockFetcher and Sequence

### LINKED ISSUE #340: CloudFlare 502 Bad Gateway errors during tests
I have discovered a wonderful new DoS against Sokol called 413 Request Entity Too Large testing.

In the fix for #254, I added a test that generated a batch that was too large by using 13001 block numbers

https://github.com/poanetwork/poa-explorer/blob/09633bde6446a640e161e9dbdf5b397436aaabc8/apps/ethereum_jsonrpc/test/etheream_jsonrpc_test.exs#L84-L109

This morning, while testing #335, I kept hitting 502 Bad Gateway errors from Cloud Flare.  I adapted #335 to handle 502 Bad Gateway, but then @phahulin informed me 

> Still, there are some disproportionally huge responses. For example, one of them weighs 130 MB and took 37 seconds to be processed. It's an array of 6537 objects.
I'm not yet sure what the request was and why response is so large - logs like this are hard to parse - need some more time to investigate, 

The only thing that requests that big of response that I can think of is the above test, so the test needs to not actually hit Sokol and instead using some mocking mechanism using [`mox`](https://hex.pm/packages/mox)

### PR DESCRIPTION
**Pull request by KronicDeth**
Resolves #340

### Enhancements
* Regression tests for #340.
* Move `chunk_ranges` from `BlockFetcher` to `Sequence` as this will guarantee that any caller of `Sequence.start_link` or `Sequence.inject_range` ends up queuing chunked ranges.
* Cap genesis sequence immediately, to indicate we don't really want infinite ranges after all the missing blocks are found, it will stop at `0` anyway.

### Bug Fixes
* Fix [off-by-one error](https://en.wikipedia.org/wiki/Off-by-one_error)s in `BlockFetcher.chunk_ranges`.
* Fix `chunk_ranges` for descending ranges.  
* Return error from `Sequence.start_link` if `:first` would cause a retraversal.  Retraversal occurs when `first` was set to the `latest_block_number` and the `step` was negative because the `prefix` of the missing block numbers which are all `<= latest_block_number` would be consumed, then the `Sequence` would start calculating infinite ranges starting `first`, which was back at `latest_block_number`.

### CODE REVIEW

**Comment by acravenho** [comment]
I started running the indexer after standup, reached block `2701794` and then the indexer started over again. Here is the error that I received:

```
[error] gensis index stream exited. Restarting
[error] Task #PID<0.26970.4> started from #PID<0.734.0> terminating
** (EthereumJSONRPC.DecodeError) Failed to decode Ethereum JSONRPC response:

  request:

    url: https://sokol.poa.network

    body: [{"id":2706313,"jsonrpc":"2.0","method":"eth_getBlockByNumber","params":["0x294B89",true]},{"id":2706312,"jsonrpc":"2.0","method":"eth_getBlockByNumber","params":["0x294B88",true]},{"id":2706311,"jsonrpc":"2.0","method":"eth_getBlockByNumber","params":["0x294B87",true]},{"id":2706310,"jsonrpc":"2.0","method":"eth_getBlockByNumber","params":["0x294B86",true]},{"id":2706309,"jsonrpc":"2.0","method":"eth_getBlockByNumber","params":["0x294B85",true]},{"id":2706308,"jsonrpc":"2.0","method":"eth_getBlockByNumber","params":["0x294B84",true]},{"id":2706307,"jsonrpc":"2.0","method":"eth_getBlockByNumber","params":["0x294B83",true]},{"id":2706306,"jsonrpc":"2.0","method":"eth_getBlockByNumber","params":["0x294B82",true]},{"id":2706305,"jsonrpc":"2.0","method":"eth_getBlockByNumber","params":["0x294B81",true]},{"id":2706304,"jsonrpc":"2.0","method":"eth_getBlockByNumber","params":["0x294B80",true]}]

  response:

    status code: 504
```

**Comment by KronicDeth** [comment]
@acravenho there are the proper 10 requests in the body of the request you posted, so that part is fixed.  If Sokol is in a timeout state, this can't fix that.  The nodes need to get into a good state first.  If code from outside this fix is still hitting Sokol, they will cause Sokol to 504.  You need to shutdown the bugged indexer to give the Parity nodes time to recover.

File: apps/indexer/test/indexer/sequence_test.exs [hunk 1]
```diff
@@ -4,15 +4,59 @@ defmodule Indexer.SequenceTest do
...
+      assert Sequence.pop(ascending) == 5..5
+      assert Sequence.pop(ascending) == 6..6
+    end
+
+    test "without :ranges with :first pop with negative step pops past 0" do
+      {:ok, descending} = Sequence.start_link(first: 1, step: -1)
+
+      assert Sequence.pop(descending) == 1..1
+      assert Sequence.pop(descending) == 0..0
+      assert Sequence.pop(descending) == -1..-1
```
Comment (tmecklem) [inline apps/indexer/test/indexer/sequence_test.exs]: Does the Sequencer go negative in practice for anything?

Comment (KronicDeth) [reply apps/indexer/test/indexer/sequence_test.exs]: No, but adding a check to prevent it made the `Sequencer` more complicated.  The last commit in the PR "Don't allow Sequence to be starts with a mix of ranges and infinite" was to make the code simpler.  It was getting really long to validate for all potential combinations that we don't actually use.

If I wanted to restrict `Sequence` to just how we use it, `ranges` should be required for `step < 0`.


.... yeah, I'll do that, it will prevent a potential class of bugs.
